hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/SimpleCopyListing.java

```diff
@@ -730,4 +660,155 @@ private void writeToFileListing(SequenceFile.Writer fileListWriter,
+                      context.getBlocksPerChunk());
+
+              for (CopyListingFileStatus fs : childCopyListingStatus) {
+                if (randomizeFileListing) {
+                  addToFileListing(fileStatuses,
+                      new FileStatusInfo(fs, sourcePathRoot), fileListWriter);
+                } else {
+                  writeToFileListing(fileListWriter, fs, sourcePathRoot);
+                }
+              }
+            }
+            if (retry < maxRetries) {
+              if (child.isDirectory()) {
+                if (LOG.isDebugEnabled()) {
+                  LOG.debug("Traversing into source dir: " + child.getPath());
+                }
+                workers.put(new WorkRequest<FileStatus>(child, retry));
+              }
+            } else {
+              LOG.error("Giving up on " + child.getPath() + " after " + retry
+                  + " retries.");
+            }
+          }
+        } catch (InterruptedException ie) {
+          LOG.error("Could not get item from childQueue. Retrying...");
+        }
+      }
+      workers.shutdown();
+    }
+
+    private void traverseDirectoryLegacy() throws IOException {
+      Stack<FileStatus> pathStack = new Stack<FileStatus>();
+      for (FileStatus fs : sourceDirs) {
+        if (excludeList == null || !excludeList
+            .contains(fs.getPath().toUri().getPath())) {
+          pathStack.add(fs);
+        }
+      }
+      while (!pathStack.isEmpty()) {
+        prepareListing(pathStack.pop().getPath());
+      }
+    }
+
+    private void prepareListing(Path path) throws IOException {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Recording source-path: " + path + " for copy.");
+      }
+      RemoteIterator<FileStatus> listStatus = sourceFS.listStatusIterator(path);
+      while (listStatus.hasNext()) {
+        FileStatus child = listStatus.next();
+        if (excludeList != null && excludeList
+            .contains(child.getPath().toUri().getPath())) {
+          continue;
+        }
+        LinkedList<CopyListingFileStatus> childCopyListingStatus = DistCpUtils
+            .toCopyListingFileStatus(sourceFS, child,
+                preserveAcls && child.isDirectory(),
+                preserveXAttrs && child.isDirectory(),
+                preserveRawXattrs && child.isDirectory(),
+                context.getBlocksPerChunk());
+        for (CopyListingFileStatus fs : childCopyListingStatus) {
+          if (randomizeFileListing) {
+            addToFileListing(fileStatuses,
+                new FileStatusInfo(fs, sourcePathRoot), fileListWriter);
+          } else {
+            writeToFileListing(fileListWriter, fs, sourcePathRoot);
+          }
+        }
+        if (child.isDirectory()) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Traversing into source dir: " + child.getPath());
+          }
+          prepareListing(child.getPath());
+        }
+      }
+    }
```

Comment:
Yeps, that is something cool, I extracted a part of it to `hadoop-common`, let me know if you have objections doing that, well I wanted to move whole of it to `common`, just left it because of the class `CallableSupplier`, I thought moving this might cause some incompatibility problems, as this was being used in the prod code as well.

Reply:
happy for you to take what's merged up. That CallableSupplier can be moved if you need to